File: Assignment 1/MergeSort/MergeSort.py
import random
import sys
from time import time


def Merge(arr,p,q,r):
    x = p
    y = r
    arr2 = arr.copy()
    for i in range(r-q):
        arr2[r-i]=arr[q+i+1]
    for i in range(r-p):
        if(arr2[x] > arr2[y]):
            arr[p+i]=arr2[x]
            x=x+1
        else:
            arr[p+i]=arr2[y]
            y=y-1
    arr[r]=arr2[x]

def MergeSort(arr,p,r):
    if(p<r):
        q = (r+p)//2
        MergeSort(arr,p,q)
        MergeSort(arr,q+1,r)
        Merge(arr,p,q,r)

#input de datos
archivo = open("input.txt").readlines()
print(archivo)
n = int(archivo[0])
datos = map(int,archivo[1].split())
arr = list(datos)

#medir tiempos
start = time()
MergeSort(arr, 0, len(arr)-1)
finish = time()

#output en un archivo
with open('output.txt', 'w') as f:
    sys.stdout = f # Change the standard output to the file we created.
    print(n)
    print(finish-start)

File: Assignment 1/MergeSort/MergeSort.cpp
#include <iostream>
using namespace std;

void Merge(int arr[],int p,int q,int r){
	int *b =new int[r-p+1];
	int i,j;
	for(i = p ; i <= q ; i++){
		b[i-p]=arr[i];
	}
	for(j = r ; j > q ; j--){
		b[i-p]=arr[j];
		i++;
	}
	i = 0;
	j = r-p;
	int k;
	for(k = p ; k < r ; k++){
		if(b[i] < b[j]){
			arr[k] = b[i];
			i++;
		}
		else{
			arr[k] = b[j];
			j--;
		}
	}
	arr[k]=b[j];
}
void MergeSort(int arr[],int p,int r){
        
        if (p < r) { 
            int q = (p + r) / 2; 
  
            MergeSort(arr, p, q); 
            MergeSort(arr, q + 1, r); 
  
            Merge(arr, p, q, r); 
        }
    }
int main(){
    
    int input,size;
    cin >> size;
    int array[size];
    for(int i = 0 ; i < size ; i++ )
    {
        cin >> input;
        array[i] = input;
    }
    
    unsigned t0,t1;

    t0 = clock();
    MergeSort(array,0,size-1);
    t1 = clock();

    double time = (double(t1-t0)/CLOCKS_PER_SEC);
    cout << time ;

    return 0;
}
